QEfficient/base/modeling_qeff.py

```diff
@@ -5,96 +5,307 @@
-"""
-
+import hashlib
+import inspect
+import json
+import logging
+import shutil
+import subprocess
+import warnings
 from abc import ABC, abstractmethod
-from enum import Enum
-from typing import Any, Dict, List, Optional
+from pathlib import Path
+from typing import Dict, List, Optional
 
-import numpy as np
+import onnx
 import torch
-from onnxruntime import InferenceSession as ORTInferenceSession
 
+from QEfficient.base.onnx_transforms import OnnxTransform
+from QEfficient.base.pytorch_transforms import PytorchTransform
 from QEfficient.generation.cloud_infer import QAICInferenceSession
+from QEfficient.utils import constants
+from QEfficient.utils.cache import QEFF_HOME, to_hashable
 
-
-# Defining placeholder ENUM for execute function
-class Runtime(Enum):
-    CPU_ORT = "CPU_ORT"
-    CPU_PT = "CPU_PT"
-    AI_100 = "AI_100"
+logger = logging.getLogger(__name__)
 
 
 class QEFFBaseModel(ABC):
     """
-    This class acts as parent class for all the varieties of model class (i.e. LLMs, SD, quantized etc.).
-    Enforces certain methods to be implemented by child classes.
+    Base class for all the model classes (i.e. LLMs, SD, quantized etc.).
+    Provides certain utility methods to be used by child classes.
 
-    All the child classes must provide way to load, transform(optimize), exoprt to ONNX etc. capabilities.
+    Class variables:
+    :_pytorch_transforms: Pytorch transformations to be applied after initialization.
+    :_onnx_transforms: ONNX transformations to be applied after ONNX export.
     """
 
+    _pytorch_transforms: List[PytorchTransform]
+    _onnx_transforms: List[OnnxTransform]
+
+    @classmethod
+    def _transform_names(cls) -> List[str]:
+        return [x.__name__ for x in cls._pytorch_transforms + cls._onnx_transforms]
+
     def __init__(self, model: torch.nn.Module) -> None:
         super().__init__()
         self.model = model
         self.onnx_path: Optional[str] = None
-        self.ort_session: Optional[ORTInferenceSession] = None
         self.qpc_path: Optional[str] = None
-        self.device_id: Optional[List[int]] = None
         self.qpc_session: Optional[QAICInferenceSession] = None
 
-    @classmethod
-    def from_pretrained(cls, pretrained_model_name_or_path: str, *args, **kwargs):
-        raise NotImplementedError("Must implement for child classes")
-
-    def run_pytorch(self, inputs):
-        raise NotImplementedError("Reached too far!!")
+        # Apply the transformations
+        any_transformed = False
+        for transform in self._pytorch_transforms:
+            self.model, transformed = transform.apply(self.model)
+            any_transformed = any_transformed or transformed
 
-    def run_ort(self, inputs: Dict[str, np.ndarray], /, cache_session: bool = False):
-        raise NotImplementedError("Reached too far!!")
+        if not any_transformed:
+            warnings.warn(f"No transforms applied to model: {self.model_name}. It may be an unsupported model!")
```

Comment:
The issue is unavoidable by us (library). But the client code can avoid it, by using a supported model. The above answer says if the client (user of the library) application should be modified, it should raise a warning. Also, if the user wants, they can capture the warnings in logging (read the second answer from the same page above).

We can do both too. Add a logging as well as warning.

Reply:
I've made it to logger warning instead of warning. I think it's still fine to provide this functionality, since we're providing clear indication that it may not work.

It may still be useful for some users. For example, we may need to use this if we have locally patched the transformers classes to QEff classes and want to use this class.